feat: add Pool type for pooling plugin instances

zshipko opened the pull request:
- `Pool` allows for a configurable number of instances for each plugin and will block waiting for a plugin to become available if the limit is reached

Nutomic commented on runtime/src/pool.rs:
```diff
@@ -0,0 +1,158 @@
+
+/// `Pool` manages threadsafe access to a limited number of instances of multiple plugins
+#[derive(Clone)]
+pub struct Pool<Key: std::fmt::Debug + Clone + std::hash::Hash + Eq = String> {
```
Im not sure what the `Key` parameter is good for. It may be useful to store multiple different plugin types in the same pool, but this is tricky as there is currently no way to get a list of all existing keys. Anyway the pool itself has very little overhead, so it makes more sense for users to write `HashMap<Key, Pool>` if needed. In my case Im using `Vec<Pool<()>>` because there is no need for any hashmap.

Nutomic commented on runtime/src/pool.rs:
```diff
@@ -0,0 +1,158 @@
+
+impl<Key: std::fmt::Debug + Clone + std::hash::Hash + Eq> Pool<Key> {
+    /// Create a new pool
+    pub fn new(max_instances: usize) -> Self {
```
In most cases this should be fine with `max_instances = std::thread::available_parallelism()`. So I would consider using that value by default, removing the parameter here and providing a builder to set a different max value.

Nutomic commented on runtime/src/pool.rs:
```diff
@@ -0,0 +1,158 @@
+            }
+            if std::time::Instant::now() - start > timeout {
+                return Ok(None);
+            }
```
The loop should sleep for 100ms or so in each iteration.

zshipko commented:
@Nutomic thanks for the review! I like the suggestion about moving the map out of `Pool` - I will take a look at making some of these changes when I get a chance.

Nutomic commented on runtime/src/pool.rs:
```diff
@@ -0,0 +1,158 @@
+
+/// `PoolPlugin` is used by the pool to track the number of live instances of a particular plugin
+#[derive(Clone, Debug)]
+pub struct PoolPlugin(std::rc::Rc<std::cell::RefCell<Plugin>>);
```
This type should probably be exported with `pub use`, at the moment its inaccessible.

Nutomic commented:
Are you planning to finish this PR and merge it? Our corresponding feature is already merged using this code as git dependency (https://github.com/LemmyNet/lemmy/pull/5498). Our stable release is still some months away, by then it would be good to depend on a stable extism version with this included.

zshipko commented:
Hi @Nutomic - I think we could release this code, especially if it's already working for you. I will try to address your feedback and get this into a mergeable state in the next few weeks.

zshipko commented:
Hi @Nutomic - I've addressed most of your feedback, with the exception of removing the `Key` type parameter. I don't think I will be able to refactor to make that change. If you really need it I would be happy to review a PR based off of this branch, otherwise I think this should be ready!
